plugins/null-ls/default.nix

```diff
@@ -12,38 +12,244 @@ in {
+
+      tempDir = helpers.defaultNullOpts.mkStr "null" ''
+        Defines the directory used to create temporary files for sources that rely on them (a
+        workaround used for command-based sources that do not support `stdio`).
+
+        To maximize compatibility, null-ls defaults to creating temp files in the same directory as
+        the parent file.
+        If this is causing issues, you can set it to `/tmp` (or another appropriate directory) here.
+        Otherwise, there is no need to change this setting.
+
+        **Note**: some null-ls built-in sources expect temp files to exist within a project for
+        context and so will not work if this option changes.
+
+        You can also configure `temp_dir` per built-in by using the `with` method, described in
+        BUILTIN_CONFIG.md.
+      '';
+
+      updateInInsert = helpers.defaultNullOpts.mkBool false ''
+        Controls whether diagnostic sources run in insert mode.
+        If set to `false`, diagnostic sources will run upon exiting insert mode, which greatly
+        improves performance but can create a slight delay before diagnostics show up.
+        Set this to `true` if you don't experience performance issues with your sources.
+
+        Note that by default, Neovim will not display updated diagnostics in insert mode.
+        Together with the option above, you need to pass `update_in_insert = true` to
+        `vim.diagnostic.config` for diagnostics to work as expected.
+        See `:help vim.diagnostic.config` for more info.
+      '';
+
+      sourcesItems = mkOption {
+        default = null;
+        # type = with types; nullOr (either (listOf str) (listOf attrsOf str));
+        type = with types; nullOr (listOf (attrsOf str));
+        description = "The list of sources to enable, should be strings of lua code. Don't use this directly";
+      };
+
+      # sources = mkOption {
+      #   default = null;
+      #   type = with types; nullOr attrs;
+      # };
     };
+
+  config = let
+    options =
+      {
+        inherit
+          (cfg)
+          border
+          cmd
+          debounce
+          debug
+          ;
+        default_timeout = cfg.defaultTimeout;
+        diagnostic_config = cfg.diagnosticConfig;
+        diagnostics_format = cfg.diagnosticsFormat;
+        fallback_severity =
+          if isString cfg.fallbackSeverity
+          then helpers.mkRaw "vim.diagnostic.severity.${strings.toUpper cfg.fallbackSeverity}"
+          else cfg.fallbackSeverity;
+        log_level = cfg.logLevel;
+        notify_format = cfg.notifyFormat;
+        on_attach =
+          helpers.ifNonNull' cfg.onAttach
+          (helpers.mkRaw cfg.onAttach);
+        on_init =
+          helpers.ifNonNull' cfg.onInit
+          (helpers.mkRaw cfg.onInit);
+        on_exit =
+          helpers.ifNonNull' cfg.onExit
+          (helpers.mkRaw cfg.onExit);
+        root_dir =
+          helpers.ifNonNull' cfg.rootDir
+          (helpers.mkRaw cfg.rootDir);
```

Comment:
We should  maybe look into a helper for this in the future

Reply:
yes it could be useful...